# Introduction
The following code visualize the [sample data](http://data.apollo.auto/help?name=data_intro_3d&data_key=lidar_obstacle_label&data_type=0&locale=en-us&lang=en) from Apollo.

Note: The visualization image will not show up on the preview of file on Github.

# Requirements

## Python packages
The following packages has to be installed.
- numpy
- numba
- scipy
- pandas
- pyntcloud
- pythreejs

All packages except pyntcloud can be install normally using `pip install`. For pyntcloud, please use

`pip install git+https://github.com/daavoo/pyntcloud`

## Data
The sample point cloud data from Apollo. The extracted folder should be at the same level as this file.

# Extra requirements for this notebook
To actually use this notebook, the following packages are also needed.
- jupyter
- pythreejs

After that, please run the following commands  
`jupyter nbextension install --py --symlink --sys-prefix pythreejs`  
`jupyter nbextension enable --py --sys-prefix pythreejs`  


In [1]:
from dataclasses import dataclass

import numpy as np
import pandas as pd
from pyntcloud import PyntCloud

In [2]:
# See https://stackoverflow.com/a/42366706/10163723
def bounding_box(points, min_x=-np.inf, max_x=np.inf, min_y=-np.inf,
                        max_y=np.inf, min_z=-np.inf, max_z=np.inf):
    """ Compute a bounding_box filter on the given points

    Parameters
    ----------                        
    points: (n,3) array
        The array containing all the points's coordinates. Expected format:
            array([
                [x1,y1,z1],
                ...,
                [xn,yn,zn]])

    min_i, max_i: float
        The bounding box limits for each coordinate. If some limits are missing,
        the default values are -infinite for the min_i and infinite for the max_i.

    Returns
    -------
    bb_filter : boolean array
        The boolean mask indicating wherever a point should be keeped or not.
        The size of the boolean mask will be the same as the number of given points.

    """

    bound_x = np.logical_and(points[:, 0] > min_x, points[:, 0] < max_x)
    bound_y = np.logical_and(points[:, 1] > min_y, points[:, 1] < max_y)
    bound_z = np.logical_and(points[:, 2] > min_z, points[:, 2] < max_z)

    bb_filter = np.logical_and(bound_x, bound_y, bound_z)

    return bb_filter

# Configuration

In [3]:
# Config dtype for numpy.
names = ('x', 'y', 'z', 'i')
offsets = (0, 4, 8, 12)
formats = ('f4', 'f4', 'f4', 'f4')
dt = np.dtype({
    'names': names,
    'offsets': offsets,
    'formats': formats,
}, align=True)

label_names = ('type', 'cx', 'cy', 'cz', 'length', 'width', 'height', 'yaw')
annotation_colors = {
    'vehicle': [255, 0, 0],
    'cyclist': [0, 255, 0],
    'pedestrian': [0, 0, 255],
    'dontCare': [255, 255, 100]
}

# Load point cloud data

In [4]:
# Load point cloud data in all frames.
point_frames = []
for i in range(100):
    frame = pd.DataFrame(np.fromfile('./3d-sample/bin_files/002_%08d.bin' % i, dt))
    
    # Give all points a white color.
    frame['red'] = np.ones(frame.shape[0]) * 255
    frame['green'] = np.ones(frame.shape[0]) * 255
    frame['blue'] = np.ones(frame.shape[0]) * 255
    
    point_frames.append(frame)

## This two code segments below process every frames, so it will take quite some time

In [5]:
@dataclass
class PointCount:
    '''Class for keeping track of point per object.'''
    obj_type: str
    frame: int
    count: int

In [6]:
pedestrian_point_count = [] # type: List[PointCount]

# Load annotation data.
label_frames = []
for i in range(100):
    frame = pd.read_csv('./3d-sample/label_file/002_%08d.bin.txt' % i, sep=' ', header=None, names=label_names)
    
    points = point_frames[i]
    for oidx, obj in frame.iterrows():
        cx = obj['cx'] 
        cy = obj['cy']
        cz = obj['cz']
        width = obj['width']
        length = obj['length']
        height = obj['height']

        min_x = cx - (0.5 * length)
        max_x = cx + (0.5 * length)
        min_y = cy - (0.5 * width)
        max_y = cy + (0.5 * width)
        min_z = cz - (0.5 * height)
        max_z = cz + (0.5 * height)

        bb = bounding_box(points.values, min_x=min_x, max_x=max_x, min_y=min_y,
                            max_y=max_y, min_z=min_z, max_z=max_z)

        #filtered_points = points.loc[bb]
        r, g, b = annotation_colors[obj['type']]
        
        # Color the points within the box with color that is corresponding to the object type.
        points.loc[bb, 'red'], points.loc[bb, 'green'], points.loc[bb, 'blue'] = r, g, b
        
        if obj['type'] == 'pedestrian':
            pc = PointCount('pedestrian', i, len(points.loc[bb, 'red'].index))
            pedestrian_point_count.append(pc)
    
    label_frames.append(frame)

# Statistic Calculation

In [7]:
point_frames[0]

x          y         z      i    red  green   blue
0      23.570656   9.279344 -2.292419  133.0  255.0  255.0  255.0
1      25.185184  11.014298 -2.352471  144.0  255.0  255.0  255.0
2      16.669695   6.983208 -2.106443  129.0  255.0  255.0  255.0
3      27.448662  13.229650 -2.436704  131.0  255.0  255.0  255.0
4      29.794174  15.765635 -2.539469  162.0  255.0  255.0  255.0
5      17.297304   8.068338 -2.140237  101.0  255.0  255.0  255.0
6      18.156616   9.279870 -2.179916  152.0  255.0  255.0  255.0
7      61.774944  19.087719 -2.283062  255.0  255.0  255.0  255.0
8      59.824184  21.004450 -1.836049  176.0  255.0  255.0  255.0
9       6.424866   1.911165 -0.281108   40.0  255.0    0.0    0.0
10     41.302471  13.826706 -2.746058   83.0  255.0  255.0  255.0
11     47.285770  17.807009 -2.934820  109.0  255.0  255.0  255.0
12     57.152683  24.029810 -3.279393  184.0  255.0  255.0  255.0
13     58.085415  26.916632 -3.032323  189.0  255.0  255.0  255.0
14     58.648823  29.983919 -2.749429  251.0  255.0  255.0  255.0
15     64.399620  18.879698  0.023242  211.0  255.0  255.0  255.0
16      4.560964   1.170064 -1.706281  129.0  255.0  255.0  255.0
17      4.599483   1.441026 -1.718071  125.0  255.0  255.0  255.0
18     10.047710   5.043281 -1.922017  114.0  255.0  255.0  255.0
19     10.538579   6.019447 -1.965521  133.0  255.0  255.0  255.0
20      4.643953   1.814432 -1.731928  151.0  255.0  255.0  255.0
21      4.655936   2.104940 -1.728920  187.0  255.0  255.0  255.0
22      3.967078   1.450536 -1.700767  235.0  255.0  255.0  255.0
23      4.030213   1.706609 -1.722236  220.0  255.0  255.0  255.0
24      4.719033   2.533561 -1.752591  112.0  255.0    0.0    0.0
25      4.711905   2.843732 -1.759519  131.0  255.0  255.0  255.0
26      3.987477   2.042457 -1.717631  132.0  255.0  255.0  255.0
27      3.980790   2.283149 -1.722757  111.0  255.0  255.0  255.0
28      5.660922   1.476202 -1.456132   61.0  255.0    0.0    0.0
29      5.992167   1.897732 -1.534506  212.0  255.0    0.0    0.0
...          ...        ...       ...    ...    ...    ...    ...
94959  10.006039   5.209896 -1.924386  109.0  255.0  255.0  255.0
94960  10.477848   6.193079 -1.966681  135.0  255.0  255.0  255.0
94961   4.616681   1.882409 -1.726949  148.0  255.0  255.0  255.0
94962   4.634568   2.177580 -1.728094  192.0  255.0  255.0  255.0
94963   3.954001   1.511722 -1.699636  231.0  255.0  255.0  255.0
94964   4.019900   1.771945 -1.724186  224.0  255.0  255.0  255.0
94965   4.691067   2.607411 -1.751626  114.0  255.0  255.0  255.0
94966   4.667920   2.910683 -1.753860  139.0  255.0  255.0  255.0
94967   3.960411   2.101742 -1.714013  155.0  255.0  255.0  255.0
94968   3.966001   2.351748 -1.726701  104.0  255.0  255.0  255.0
94969   5.731284   1.585409 -1.478158   64.0  255.0    0.0    0.0
94970   6.135218   2.044387 -1.578658  193.0  255.0    0.0    0.0
94971   5.506477   1.380052 -1.744829  121.0  255.0    0.0    0.0
94972   5.610665   1.720642 -1.750086  103.0  255.0    0.0    0.0
94973   7.005402   2.853854 -1.812083  144.0  255.0  255.0  255.0
94974   7.111310   3.338271 -1.814583  130.0  255.0  255.0  255.0
94975   5.609986   2.153073 -1.752001  100.0  255.0    0.0    0.0
94976   5.693944   2.519985 -1.765034  134.0  255.0    0.0    0.0
94977   7.195561   3.956890 -1.817804  109.0  255.0  255.0  255.0
94978   7.483470   4.644745 -1.860728  139.0  255.0  255.0  255.0
94979   5.765521   3.023071 -1.785074  108.0  255.0  255.0  255.0
94980   5.806485   3.444119 -1.790850  126.0  255.0  255.0  255.0
94981   5.819551   1.635965 -0.826389  180.0  255.0    0.0    0.0
94982   6.139857   2.044599 -0.867780  215.0  255.0    0.0    0.0
94983   5.593024   1.431944 -1.094272  145.0  255.0    0.0    0.0
94984   5.895759   1.822819 -1.151577  215.0  255.0    0.0    0.0
94985   6.825604   2.775352 -0.948687  229.0  255.0  255.0  255.0
94986   7.512899   3.502051 -1.000795   81.0  255.0  255.0  255.0
94987   6.526774   2.498386 -1.264912  219.0  255.0    

Average number of points in a frame is

In [8]:
points_in_frame = np.array([len(f.index) for f in point_frames])
np.average(points_in_frame)

99345.93

In [9]:
pedestrian_point_count

[PointCount(obj_type='pedestrian', frame=0, count=19),
 PointCount(obj_type='pedestrian', frame=0, count=18),
 PointCount(obj_type='pedestrian', frame=0, count=7),
 PointCount(obj_type='pedestrian', frame=0, count=8),
 PointCount(obj_type='pedestrian', frame=0, count=12),
 PointCount(obj_type='pedestrian', frame=1, count=13),
 PointCount(obj_type='pedestrian', frame=1, count=7),
 PointCount(obj_type='pedestrian', frame=1, count=5),
 PointCount(obj_type='pedestrian', frame=2, count=30),
 PointCount(obj_type='pedestrian', frame=2, count=39),
 PointCount(obj_type='pedestrian', frame=2, count=10),
 PointCount(obj_type='pedestrian', frame=4, count=132),
 PointCount(obj_type='pedestrian', frame=4, count=20),
 PointCount(obj_type='pedestrian', frame=5, count=21),
 PointCount(obj_type='pedestrian', frame=5, count=24),
 PointCount(obj_type='pedestrian', frame=6, count=27),
 PointCount(obj_type='pedestrian', frame=7, count=16),
 PointCount(obj_type='pedestrian', frame=7, count=3),
 PointCount(ob

In [10]:
pedestrian_point_counts = np.array([c.count for c in pedestrian_point_count])

In [11]:
len(pedestrian_point_count)

396

Thus, in these 100 frames, there are 396 objects that are labeled as pedestrian.

The average of number of points for each pedestrian is

In [12]:
pedestrian_point_average = np.sum(pedestrian_point_counts) / len(pedestrian_point_count)
pedestrian_point_average

62.89141414141414

In [13]:
print('Min: %.3f' % np.min(pedestrian_point_counts))
print('Max: %.3f' % np.max(pedestrian_point_counts))

Min: 1.000
Max: 959.000


# Displaying the points

- vehicle = red
- cyclist = green
- pedestrian = blue
- dontCare = yellow

In [14]:
points = point_frames[15] # Selecting a frame.

In [15]:
cloud = PyntCloud(points)

Showing the rendered point could. This will not show up in a preview of file in Github website.

In [16]:
cloud.plot(initial_point_size=0.01, backend='pythreejs')

Renderer(camera=PerspectiveCamera(aspect=1.6, fov=90.0, position=(0.1779005378484726, 53.93962222337723, 4.133…